icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameMultiplexer.java

```diff
@@ -0,0 +1,104 @@
+    private List<MixedUnitLongNameHandler> fMixedUnitHandlers;
+
+    // Unowned pointers to instances owned by MaybeStackVectors.
+    private List<MicroPropsGenerator> fHandlers;
+
+    // Each MeasureUnit corresponds to the same-index MicroPropsGenerator
+    // pointed to in fHandlers.
+    private List<MeasureUnit> fMeasureUnits;
+
+    public LongNameMultiplexer(MicroPropsGenerator fParent) {
+        this.fParent = fParent;
+    }
+
+    // Produces a multiplexer for LongNameHandlers, one for each unit in
+    // `units`. An individual unit might be a mixed unit.
+    public static LongNameMultiplexer forMeasureUnits(ULocale locale,
+                                                      List<MeasureUnit> units,
+                                                      NumberFormatter.UnitWidth width,
+                                                      PluralRules rules,
+                                                      MicroPropsGenerator parent) {
+        LongNameMultiplexer result = new LongNameMultiplexer(parent);
+
+        assert (units.size() > 0);
+
+        result.fMeasureUnits = new ArrayList<>();
+        result.fHandlers = new ArrayList<>();
+        result.fMixedUnitHandlers = new ArrayList<>();
+        result.fLongNameHandlers = new ArrayList<>();
+
+        for (int i = 0; i < units.size(); i++) {
+            MeasureUnit unit = units.get(i);
+            result.fMeasureUnits.add(unit);
+            if (unit.getComplexity() == MeasureUnit.Complexity.MIXED) {
+                MixedUnitLongNameHandler mlnh = MixedUnitLongNameHandler
+                        .forMeasureUnit(locale, unit, width, rules, null); /*TODO: why is null*/
+                result.fMixedUnitHandlers.add(mlnh);
+                result.fHandlers.add(mlnh);
+            } else {
+                LongNameHandler lnh = LongNameHandler
+                        .forMeasureUnit(locale, unit, NoUnit.BASE, width, rules, null ); /*TODO: why is null */
+                result.fLongNameHandlers.add(lnh);
+                result.fHandlers.add(lnh);
+            }
+        }
+
+        return result;
+    }
+
+    // The output unit must be provided via `micros.outputUnit`, it must match
+    // one of the units provided to the factory function.
+    @Override
+    public MicroProps processQuantity(DecimalQuantity quantity) {
+
+        // We call parent->processQuantity() from the Multiplexer, instead of
+        // letting LongNameHandler handle it: we don't know which LongNameHandler to
+        // call until we've called the parent!
+        MicroProps micros = this.fParent.processQuantity(quantity);
+
+        // Call the correct LongNameHandler based on outputUnit
+        for (int i = 0; i < this.fHandlers.size(); i++) {
+            if (fMeasureUnits.get(i).equals( micros.outputUnit)) {
+                // FIXME: ugly workaround
+                MicroPropsGenerator h = fHandlers.get(i);
+                if (MixedUnitLongNameHandler.class.isInstance(h)) {
```

Comment:
```java
if (h instanceof MixedUnitLongNameHandler) {
  MixedUnitLongNameHandler hh = (MixedUnitLongNameHandler) h;
  return hh.processQuantityWithMicros(quantity, micros);
}
```

and likewise below with LongNameHandler.

Alternatively, you could make an interface with the `processQuantityWithMicros` function on it, and both classes (MixedUnitLNH and LNH) can implement that interface.

Reply:
Agreed on the interface idea: when I wrote this "isinstance" hack, I also thought an interface is probably the right solution.

Reply:
done ... I just used the hack, we can make the interface later, even I prefer the interface idea more

Reply:
I doubt we'd get back to doing the interface later if we don't do it now. So I did it now with 15e46ba6cd.